### PR TITLE
Stop flushing fpc when not asked for

### LINKED ISSUE #38301: [Issue] Stop flushing fpc when not asked for
This issue is automatically created based on existing pull request: magento/magento2#38215: Stop flushing fpc when not asked for

---------
### Description (*)
We found out by surprise today on Slack (thanks @benjamin-volle & @robinero) that running `bin/magento cache:clean` or `bin/magento cache:flush` will always flush FPC (built-in or Varnish) entirely, even if you've only provided a type to be cleaned like `config`, `layout`, ...

I understand that yes, probably if one of those types get flushed, you'll most likely want to flush FPC as well, but let that be up to the user to decide by seeing if they provided the type `full_page` explicitly in the command.

This fixes this.

Some history of this problem, flushing FPC via CLI got introduced at the request of https://github.com/magento/magento2/issues/2025 and got fixed by https://github.com/magento/magento2/commit/c75c203eb96d48b45d2f877d29253b8cb6067f7e

This PR enhances this a bit and will only dispatch those 2 events when either no types were provided in the command line or when `full_page` is one of the types provided.

### Fixed Issues (if relevant)
No idea

### Manual testing scenarios (*)
1. Modify this code like so:
```diff
diff --git a/app/code/Magento/CacheInvalidate/Observer/FlushAllCacheObserver.php b/app/code/Magento/CacheInvalidate/Observer/FlushAllCacheObserver.php
index 574263ca1ee..1db01887467 100644
--- a/app/code/Magento/CacheInvalidate/Observer/FlushAllCacheObserver.php
+++ b/app/code/Magento/CacheInvalidate/Observer/FlushAllCacheObserver.php
@@ -45,6 +45,8 @@ class FlushAllCacheObserver implements ObserverInterface
      */
     public function execute(\Magento\Framework\Event\Observer $observer)
     {
+        die(sprintf("Flusing FPC in %s\n", $this::class));
+
         if ($this->config->getType() == \Magento\PageCache\Model\Config::VARNISH && $this->config->isEnabled()) {
             $this->purgeCache->sendPurgeRequest(['.*']);
         }
diff --git a/app/code/Magento/PageCache/Observer/FlushAllCache.php b/app/code/Magento/PageCache/Observer/FlushAllCache.php
index 6bcfd0e38df..d7684e25d4c 100644
--- a/app/code/Magento/PageCache/Observer/FlushAllCache.php
+++ b/app/code/Magento/PageCache/Observer/FlushAllCache.php
@@ -62,6 +62,8 @@ class FlushAllCache implements ObserverInterface
      */
     public function execute(Observer $observer)
     {
+        die(sprintf("Flushing FPC in %s\n", $this::class));
+
         if ($this->_config->getType() === Config::BUILT_IN) {
             $this->fullPageCache->clean();
         }
```

2. Use built-in FPC
3. Run `bin/magento cache:flush` and `bin/magento cache:clean` => expected to hit one of those `die` statements introduced in step 1
4. Run `bin/magento cache:flush full_page` and `bin/magento cache:clean full_page` => expected to hit one of those `die` statements introduced in step 1
5. Run `bin/magento cache:flush layout` and `bin/magento cache:clean layout` => expected to **not** hit one of those `die` statements introduced in step 1
6. Configure FPC to use Varnish
7. Run steps 3 to 5 again

### Questions or comments
<del>Since we introduced `Magento\PageCache` in the `Magento_Backend` module, should we add `magento/module-page-cache` as dependency in the Backend module's `composer.json` file? We can also choose to hardcode `full_page` and don't use the constant, that's also an option ...</del>
_Update_: looks like the automated tests also picked this up, I already force-pushed a change to have it hardcoded instead of using a constant from another module, adding a dependency might be out of scope here.

I noticed that the New Relic module also [listens for that `adminhtml_cache_flush_system` event](https://github.com/magento/magento2/blob/298cbd43f6effcc18b554bf1fc8cf6608480077f/app/code/Magento/NewRelicReporting/etc/adminhtml/events.xml#L9-L12), but it's done in the `adminhtml/events.xml` file, so that should be fine since it's in the `adminhtml` area. Running those cli commands wouldn't have triggered this observer anyways. So no regression there.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by hostep**
### Description (*)
We found out by surprise today on Slack (thanks @benjamin-volle & @robinero) that running `bin/magento cache:clean` or `bin/magento cache:flush` will always flush FPC (built-in or Varnish) entirely, even if you've only provided a type to be cleaned like `config`, `layout`, ...

I understand that yes, probably if one of those types get flushed, you'll most likely want to flush FPC as well, but let that be up to the user to decide by seeing if they provided the type `full_page` explicitly in the command.

This fixes this.

Some history of this problem, flushing FPC via CLI got introduced at the request of https://github.com/magento/magento2/issues/2025 and got fixed by https://github.com/magento/magento2/commit/c75c203eb96d48b45d2f877d29253b8cb6067f7e

This PR enhances this a bit and will only dispatch those 2 events when either no types were provided in the command line or when `full_page` is one of the types provided.

### Fixed Issues (if relevant)
No idea

### Manual testing scenarios (*)
1. Modify this code like so:
```diff
diff --git a/app/code/Magento/CacheInvalidate/Observer/FlushAllCacheObserver.php b/app/code/Magento/CacheInvalidate/Observer/FlushAllCacheObserver.php
index 574263ca1ee..1db01887467 100644
--- a/app/code/Magento/CacheInvalidate/Observer/FlushAllCacheObserver.php
+++ b/app/code/Magento/CacheInvalidate/Observer/FlushAllCacheObserver.php
@@ -45,6 +45,8 @@ class FlushAllCacheObserver implements ObserverInterface
      */
     public function execute(\Magento\Framework\Event\Observer $observer)
     {
+        die(sprintf("Flusing FPC in %s\n", $this::class));
+
         if ($this->config->getType() == \Magento\PageCache\Model\Config::VARNISH && $this->config->isEnabled()) {
             $this->purgeCache->sendPurgeRequest(['.*']);
         }
diff --git a/app/code/Magento/PageCache/Observer/FlushAllCache.php b/app/code/Magento/PageCache/Observer/FlushAllCache.php
index 6bcfd0e38df..d7684e25d4c 100644
--- a/app/code/Magento/PageCache/Observer/FlushAllCache.php
+++ b/app/code/Magento/PageCache/Observer/FlushAllCache.php
@@ -62,6 +62,8 @@ class FlushAllCache implements ObserverInterface
      */
     public function execute(Observer $observer)
     {
+        die(sprintf("Flushing FPC in %s\n", $this::class));
+
         if ($this->_config->getType() === Config::BUILT_IN) {
             $this->fullPageCache->clean();
         }
```

2. Use built-in FPC
3. Run `bin/magento cache:flush` and `bin/magento cache:clean` => expected to hit one of those `die` statements introduced in step 1
4. Run `bin/magento cache:flush full_page` and `bin/magento cache:clean full_page` => expected to hit one of those `die` statements introduced in step 1
5. Run `bin/magento cache:flush layout` and `bin/magento cache:clean layout` => expected to **not** hit one of those `die` statements introduced in step 1
6. Configure FPC to use Varnish
7. Run steps 3 to 5 again

### Questions or comments
<del>Since we introduced `Magento\PageCache` in the `Magento_Backend` module, should we add `magento/module-page-cache` as dependency in the Backend module's `composer.json` file? We can also choose to hardcode `full_page` and don't use the constant, that's also an option ...</del>
_Update_: looks like the automated tests also picked this up, I already force-pushed a change to have it hardcoded instead of using a constant from another module, adding a dependency might be out of scope here.

I noticed that the New Relic module also [listens for that `adminhtml_cache_flush_system` event](https://github.com/magento/magento2/blob/298cbd43f6effcc18b554bf1fc8cf6608480077f/app/code/Magento/NewRelicReporting/etc/adminhtml/events.xml#L9-L12), but it's done in the `adminhtml/events.xml` file, so that should be fine since it's in the `adminhtml` area. Running those cli commands wouldn't have triggered this observer anyways. So no regression there.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#38301: Stop flushing fpc when not asked for

### CODE REVIEW

**Comment by hostep** [comment]
@magento run all tests

**Comment by hostep** [comment]
@magento run all tests

File: app/code/Magento/Backend/Console/Command/CacheCleanCommand.php [hunk 1]
```diff
@@ -32,12 +32,15 @@ protected function configure()
      */
     protected function performAction(array $cacheTypes)
     {
-        $this->eventManager->dispatch('adminhtml_cache_flush_system');
+        if ($cacheTypes === [] || in_array('full_page', $cacheTypes)) {
```
Comment (benjamin-volle) [inline app/code/Magento/Backend/Console/Command/CacheCleanCommand.php:35]: I believe you can use \Magento\PageCache\Model\Cache\Type::TYPE_IDENTIFIER here

Comment (hostep) [reply app/code/Magento/Backend/Console/Command/CacheCleanCommand.php:35]: I know, and that's what I did first: https://github.com/magento/magento2/compare/01b8366baffebebb72e206fa2fff54ec384de08d..53b87ab003e20974497f3d450606030a93edd335 ;)

See my remark in the description, it requires adding a composer dependency between those two modules, and I'm not sure if that's worth it in this case...

Comment (benjamin-volle) [reply app/code/Magento/Backend/Console/Command/CacheCleanCommand.php:35]: Oh, you're right, and to be honest I'm pretty sure there is approximately 0 chance on 100 that this value could change any day

**Review by ihor-sviziev — APPROVED** [review]

**Comment by engcom-Echo** [comment]
@magento create issue

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
### :heavy_check_mark: QA Passed

**Preconditions:**
- Install fresh Magento **`2.4-develop`**

**Manual testing scenario:**

1. We have follow steps from [comment](https://github.com/magento/magento2/pull/38215#issue-2008654786)


**Before:** :heavy_multiplication_x: 
Screenshot 2023-12-26 at 10 55 57 AM


**After:** :heavy_check_mark:  
Screenshot 2023-12-26 at 10 57 50 AM

**Comment by engcom-Echo** [comment]
@magento run Integration Tests,Functional Tests EE,Functional Tests CE,Functional Tests B2B

**Comment by engcom-Echo** [comment]
Functional Tests B2B and Functional Tests CE failure are different on last two run on same commit.
Functional Tests CE
Run1
Screenshot 2023-12-26 at 4 36 11 PM

Run2
Screenshot 2023-12-26 at 4 36 26 PM

Functional Tests B2B
Run1
Screenshot 2023-12-26 at 4 36 32 PM

Run2
Screenshot 2023-12-26 at 4 36 16 PM

Hence Moving it To `merge in progress`

**Comment by engcom-Charlie** [comment]
@magento run Semantic Version Checker

**Pull request merged**